tilavarauspalvelu/models/user/actions.py

```diff
@@ -1,11 +1,164 @@
 from __future__ import annotations
 
+import uuid
 from typing import TYPE_CHECKING
 
+from auditlog.models import LogEntry
+from dateutil.relativedelta import relativedelta
+from django.utils import timezone
+from social_django.models import UserSocialAuth
+
+from tilavarauspalvelu.enums import (
+    ApplicationStatusChoice,
+    OrderStatus,
+    ReservationNotification,
+    ReservationStateChoice,
+    ReservationTypeChoice,
+)
+from tilavarauspalvelu.models import (
+    Address,
+    Application,
+    ApplicationSection,
+    GeneralRole,
+    Person,
+    Reservation,
+    UnitRole,
+)
+from tilavarauspalvelu.typing import UserAnonymizationInfo
+
 if TYPE_CHECKING:
     from .model import User
 
 
+ANONYMIZED_FIRST_NAME = "ANON"
+ANONYMIZED_LAST_NAME = "ANONYMIZED"
+ANONYMIZED = "Anonymized"
+SENSITIVE_RESERVATION = "Sensitive data of this reservation has been anonymized by a script"
+SENSITIVE_APPLICATION = "Sensitive data of this application has been anonymized by a script"
+
+
 class UserActions:
     def __init__(self, user: User) -> None:
         self.user = user
+
+    def anonymize(self) -> None:
+        self.anonymize_user()
+        self.anonymize_user_applications()
+        self.anonymize_user_reservations()
+
+    def anonymize_user(self) -> None:
+        self.user.first_name = ANONYMIZED_FIRST_NAME
+        self.user.last_name = ANONYMIZED_LAST_NAME
+        self.user.email = f"{self.user.first_name}.{self.user.last_name}@anonymized.net"
+        self.user.uuid = uuid.uuid4()
+        self.user.username = f"anonymized-{self.user.uuid}"
+        self.user.date_of_birth = None
+        self.user.reservation_notification = ReservationNotification.NONE
+        self.user.is_active = False
+        self.user.is_superuser = False
+        self.user.is_staff = False
+        self.user.profile_id = ""
+        self.user.save()
+
+        self.user.ad_groups.clear()
+
+        GeneralRole.objects.filter(user=self.user).delete()
+        UnitRole.objects.filter(user=self.user).delete()
+        UserSocialAuth.objects.filter(user=self.user).delete()
+
+    def anonymize_user_reservations(self) -> None:
+        reservations = Reservation.objects.filter(user=self.user).exclude(
+            type__in=[ReservationTypeChoice.BLOCKED, ReservationTypeChoice.STAFF],
+        )
+        for reservation in reservations:
```

Comment:
Strictly speaking not part of this ticket spec. but if these (filtered) reservations belong to a recurring reservation, then the name & description strings of those parents should also be anonymized. (Not previously thought of since `SEASONAL` reservations did not exist.)

Reply:
Added anonymization for these, and made existing anonymization more performant by using `bulk_update`.